src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+interface ToFieldElements<Field, Type> {
+  toFieldElements(this: Type): Field[];
+}
+
+interface OfFieldElements<Field, Type> {
+  ofFieldElements(elts: Field[]): Type;
+}
+
+interface AsFieldElements<Field, Type>
+  extends ToFieldElements<Field, Type>,
+    OfFieldElements<Field, Type> {}
+
+/* NB: Do not combine this with other types using |, otherwise the error
+ * message is an unhelpful 'cannot convert to string'.
+ */
+type FieldConvertible = number | string | boolean;
```

Comment:
`FieldPrimitive` of `PrimitiveValue` might be better names for this type.

Reply:
I'm not very well-versed in typescript. so this next suggestion might just not make sense.

Pretty much all the functions that accept `FieldConvertible` for an argument also accept some locally scoped `Field` type parameter. It might be cleaner to rewrite this type as `type FieldConvertible<Field> = number | string | boolean | AsFieldElements<Field, _>` (though I'm not sure if there is a way to existentially quantify the `Type` variable in this scope like this).

Reply:
Yea, this is exactly what I originally wrote. Unfortunately, doing it this way means that any type errors fall back to an attempt to convert to string, whereas if we split the types you get an error that offers you the full menu of types. If there's a way around this, I'll gladly take it!